File: clientes/javascript/cliente_remoto.js
/**
 * Cliente Remoto Smart Home - JavaScript/Node.js
 * Comunica-se com o servidor Java via API REST HTTP usando JSON.
 */

const http = require('http');

class ClienteRemotoAPI {
    /**
     * Inicializa o cliente.
     * @param {string} host - Endereço do servidor
     * @param {number} porta - Porta do servidor
     */
    constructor(host = 'localhost', porta = 8080) {
        this.baseUrl = `http://${host}:${porta}`;
    }

    /**
     * Executa uma requisição HTTP e retorna a resposta.
     * @param {string} endpoint - Endpoint da API
     * @param {string} method - Método HTTP (GET, POST)
     * @param {any} dados - Dados para enviar no corpo da requisição
     * @returns {Promise<any>} Resposta deserializada
     */
    async _fazerRequisicao(endpoint, method = 'GET', dados = null) {
        return new Promise((resolve, reject) => {
            const url = new URL(this.baseUrl + endpoint);
            
            const options = {
                hostname: url.hostname,
                port: url.port,
                path: url.pathname,
                method: method,
                headers: {}
            };

            // Sempre solicita JSON
            options.headers['Accept'] = 'application/json';
            
            if (dados !== null) {
                const jsonData = JSON.stringify(dados);
                options.headers['Content-Type'] = 'application/json';
                options.headers['Content-Length'] = Buffer.byteLength(jsonData);
            }

            console.log(`[CLIENTE] ${method} ${endpoint}`);

            const req = http.request(options, (res) => {
                let responseData = '';

                res.on('data', (chunk) => {
                    responseData += chunk;
                });

                res.on('end', () => {
                    try {
                        // Tenta deserializar como JSON
                        const parsed = JSON.parse(responseData);
                        resolve(parsed);
                    } catch (e) {
                        // Se não for JSON, retorna como string
                        resolve(responseData);
                    }
                });
            });

            req.on('error', (error) => {
                reject(new Error(`Erro na requisição: ${error.message}`));
            });

            if (dados !== null) {
                req.write(JSON.stringify(dados));
            }

            req.end();
        });
    }

    /**
     * Lista todos os dispositivos IoT.
     * @returns {Promise<Array>} Lista de dispositivos
     */
    async listarDispositivos() {
        return this._fazerRequisicao('/api/dispositivos', 'GET');
    }

    /**
     * Obtém um dispositivo por ID.
     * @param {string} dispositivoId - ID do dispositivo
     * @returns {Promise<Object>} Dispositivo encontrado
     */
    async obterDispositivo(dispositivoId) {
        return this._fazerRequisicao('/api/dispositivos/obter', 'POST', dispositivoId);
    }

    /**
     * Atualiza um dispositivo.
     * @param {string} dispositivoId - ID do dispositivo
     * @param {Object} dispositivo - Dados do dispositivo
     * @returns {Promise<Object>} Dispositivo atualizado
     */
    async atualizarDispositivo(dispositivoId, dispositivo) {
        const dados = [dispositivoId, dispositivo];
        return this._fazerRequisicao('/api/dispositivos/atualizar', 'POST', dados);
    }

    /**
     * Executa uma ação em um dispositivo.
     * @param {string} dispositivoId - ID do dispositivo
     * @param {string} comando - Comando a executar
     * @returns {Promise<Object>} Dispositivo após execução
     */
    async executarAcao(dispositivoId, comando) {
        const dados = [dispositivoId, comando];
        return this._fazerRequisicao('/api/dispositivos/acao', 'POST', dados);
    }

    /**
     * Lista todas as rotinas.
     * @returns {Promise<Array>} Lista de rotinas
     */
    async listarRotinas() {
        return this._fazerRequisicao('/api/rotinas', 'GET');
    }

    /**
     * Cria uma nova rotina.
     * @param {Object} rotina - Dados da rotina
     * @returns {Promise<Object>} Rotina criada
     */
    async criarRotina(rotina) {
        return this._fazerRequisicao('/api/rotinas/criar', 'POST', rotina);
    }

    /**
     * Lista todos os alertas.
     * @returns {Promise<Array>} Lista de alertas
     */
    async listarAlertas() {
        return this._fazerRequisicao('/api/alertas', 'GET');
    }

    /**
     * Obtém um cômodo por nome.
     * @param {string} nomeComodo - Nome do cômodo
     * @returns {Promise<Object>} Cômodo encontrado
     */
    async obterComodo(nomeComodo) {
        return this._fazerRequisicao('/api/comodos/obter', 'POST', nomeComodo);
    }
}

const readline = require('readline');

/**
 * Cria interface de leitura do terminal.
 */
function criarInterface() {
    return readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
}

/**
 * Faz uma pergunta ao usuário.
 */
function perguntar(rl, pergunta) {
    return new Promise((resolve) => {
        rl.question(pergunta, (resposta) => {
            resolve(resposta);
        });
    });
}

/**
 * Exibe o menu principal.
 */
function exibirMenu() {
    console.log('\n' + '='.repeat(40));
    console.log('MENU PRINCIPAL');
    console.log('='.repeat(40));
    console.log('1. Listar dispositivos');
    console.log('2. Obter dispositivo por ID');
    console.log('3. Atualizar dispositivo');
    console.log('4. Executar acao em dispositivo');
    console.log('5. Listar rotinas');
    console.log('6. Criar rotina');
    console.log('7. Listar alertas');
    console.log('8. Obter comodo por nome');
    console.log('0. Sair');
    console.log('='.repeat(40));
}

/**
 * Lista todos os dispositivos.
 */
async function listarDispositivosInterativo(cliente) {
    console.log('\n--- Listando Dispositivos ---');
    try {
        const dispositivos = await cliente.listarDispositivos();
        console.log(`Total: ${dispositivos.length} dispositivos\n`);
        
        if (dispositivos.length === 0) {
            console.log('Nenhum dispositivo encontrado.');
        } else {
            dispositivos.forEach((d, i) => {
                console.log(`${i + 1}. ${d.nome || 'N/A'}`);
                console.log(`   ID: ${d.id || 'N/A'}`);
                console.log(`   Tipo: ${d.tipo || 'N/A'}`);
                console.log(`   Online: ${d.online ? 'Sim' : 'Nao'}`);
                console.log(`   Comodo: ${d.comodo || 'N/A'}`);
                console.log();
            });
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Obtém um dispositivo por ID.
 */
async function obterDispositivoInterativo(cliente, rl) {
    console.log('\n--- Obter Dispositivo ---');
    const dispositivoId = await perguntar(rl, 'Digite o ID do dispositivo: ');
    
    if (!dispositivoId.trim()) {
        console.log('[ERRO] ID nao pode ser vazio!');
        return;
    }
    
    try {
        const dispositivo = await cliente.obterDispositivo(dispositivoId.trim());
        if (dispositivo) {
            console.log('\nDispositivo encontrado:');
            console.log(`  Nome: ${dispositivo.nome || 'N/A'}`);
            console.log(`  ID: ${dispositivo.id || 'N/A'}`);
            console.log(`  Tipo: ${dispositivo.tipo || 'N/A'}`);
            console.log(`  Online: ${dispositivo.online ? 'Sim' : 'Nao'}`);
            console.log(`  Comodo: ${dispositivo.comodo || 'N/A'}`);
        } else {
            console.log('\n[ERRO] Dispositivo nao encontrado!');
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Atualiza um dispositivo.
 */
async function atualizarDispositivoInterativo(cliente, rl) {
    console.log('\n--- Atualizar Dispositivo ---');
    const dispositivoId = await perguntar(rl, 'Digite o ID do dispositivo: ');
    
    if (!dispositivoId.trim()) {
        console.log('[ERRO] ID nao pode ser vazio!');
        return;
    }
    
    try {
        const dispositivo = await cliente.obterDispositivo(dispositivoId.trim());
        if (!dispositivo) {
            console.log('[ERRO] Dispositivo nao encontrado!');
            return;
        }
        
        console.log('\nDispositivo atual:');
        console.log(`  Nome: ${dispositivo.nome || 'N/A'}`);
        console.log(`  Online: ${dispositivo.online ? 'Sim' : 'Nao'}`);
        
        const novoNome = await perguntar(rl, '\nNovo nome (Enter para manter): ');
        if (novoNome.trim()) {
            dispositivo.nome = novoNome.trim();
        }
        
        const onlineStr = await perguntar(rl, 'Online? (s/n, Enter para manter): ');
        if (onlineStr.trim().toLowerCase() === 's') {
            dispositivo.online = true;
        } else if (onlineStr.trim().toLowerCase() === 'n') {
            dispositivo.online = false;
        }
        
        const atualizado = await cliente.atualizarDispositivo(dispositivoId.trim(), dispositivo);
        if (atualizado) {
            console.log('\n[SUCESSO] Dispositivo atualizado!');
            console.log(`  Nome: ${atualizado.nome || 'N/A'}`);
            console.log(`  Online: ${atualizado.online ? 'Sim' : 'Nao'}`);
        } else {
            console.log('\n[ERRO] Falha ao atualizar dispositivo!');
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Executa uma ação em um dispositivo.
 */
async function executarAcaoInterativo(cliente, rl) {
    console.log('\n--- Executar Acao ---');
    const dispositivoId = await perguntar(rl, 'Digite o ID do dispositivo: ');
    
    if (!dispositivoId.trim()) {
        console.log('[ERRO] ID nao pode ser vazio!');
        return;
    }
    
    const comando = await perguntar(rl, 'Digite o comando (ligar/desligar): ');
    
    if (!comando.trim()) {
        console.log('[ERRO] Comando nao pode ser vazio!');
        return;
    }
    
    try {
        const resultado = await cliente.executarAcao(dispositivoId.trim(), comando.trim());
        if (resultado) {
            console.log('\n[SUCESSO] Acao executada!');
            console.log(`  Dispositivo: ${resultado.nome || 'N/A'}`);
            console.log(`  Online: ${resultado.online ? 'Sim' : 'Nao'}`);
        } else {
            console.log('\n[ERRO] Falha ao executar acao!');
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Lista todas as rotinas.
 */
async function listarRotinasInterativo(cliente) {
    console.log('\n--- Listando Rotinas ---');
    try {
        const rotinas = await cliente.listarRotinas();
        console.log(`Total: ${rotinas.length} rotinas\n`);
        
        if (rotinas.length === 0) {
            console.log('Nenhuma rotina encontrada.');
        } else {
            rotinas.forEach((r, i) => {
                console.log(`${i + 1}. ${r.nome || 'N/A'}`);
                console.log(`   ID: ${r.id || 'N/A'}`);
                console.log(`   Acoes: ${r.acoes ? r.acoes.length : 0}`);
                console.log();
            });
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Cria uma nova rotina.
 */
async function criarRotinaInterativo(cliente, rl) {
    console.log('\n--- Criar Rotina ---');
    const nome = await perguntar(rl, 'Nome da rotina: ');
    
    if (!nome.trim()) {
        console.log('[ERRO] Nome nao pode ser vazio!');
        return;
    }
    
    try {
        const dispositivos = await cliente.listarDispositivos();
        if (dispositivos.length === 0) {
            console.log('[ERRO] Nenhum dispositivo disponivel!');
            return;
        }
        
        console.log('\nDispositivos disponiveis:');
        dispositivos.forEach((d, i) => {
            console.log(`${i + 1}. ${d.nome || 'N/A'} (${d.id || 'N/A'})`);
        });
        
        const escolha = await perguntar(rl, '\nEscolha o numero do dispositivo: ');
        const idx = parseInt(escolha) - 1;
        
        if (isNaN(idx) || idx < 0 || idx >= dispositivos.length) {
            console.log('[ERRO] Escolha invalida!');
            return;
        }
        
        const comando = await perguntar(rl, 'Comando (ligar/desligar): ');
        
        const rotina = {
            nome: nome.trim(),
            acoes: [{
                dispositivoId: dispositivos[idx].id,
                comando: comando.trim(),
                parametros: {}
            }]
        };
        
        const criada = await cliente.criarRotina(rotina);
        if (criada) {
            console.log('\n[SUCESSO] Rotina criada!');
            console.log(`  Nome: ${criada.nome || 'N/A'}`);
            console.log(`  ID: ${criada.id || 'N/A'}`);
        } else {
            console.log('\n[ERRO] Falha ao criar rotina!');
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Lista todos os alertas.
 */
async function listarAlertasInterativo(cliente) {
    console.log('\n--- Listando Alertas ---');
    try {
        const alertas = await cliente.listarAlertas();
        console.log(`Total: ${alertas.length} alertas\n`);
        
        if (alertas.length === 0) {
            console.log('Nenhum alerta encontrado.');
        } else {
            alertas.forEach((a, i) => {
                console.log(`${i + 1}. ${a.titulo || 'N/A'}`);
                console.log(`   Mensagem: ${a.mensagem || 'N/A'}`);
                console.log(`   Comodo: ${a.comodo || 'N/A'}`);
                console.log();
            });
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Obtém um cômodo por nome.
 */
async function obterComodoInterativo(cliente, rl) {
    console.log('\n--- Obter Comodo ---');
    const nome = await perguntar(rl, 'Digite o nome do comodo: ');
    
    if (!nome.trim()) {
        console.log('[ERRO] Nome nao pode ser vazio!');
        return;
    }
    
    try {
        const comodo = await cliente.obterComodo(nome.trim());
        if (comodo) {
            console.log('\nComodo encontrado:');
            console.log(`  Nome: ${comodo.nome || 'N/A'}`);
            if (comodo.dispositivos) {
                console.log(`  Dispositivos: ${comodo.dispositivos.length}`);
                comodo.dispositivos.forEach(d => {
                    console.log(`    - ${d.nome || 'N/A'}`);
                });
            }
        } else {
            console.log('\n[ERRO] Comodo nao encontrado!');
        }
    } catch (error) {
        console.log(`[ERRO] ${error.message}`);
    }
}

/**
 * Função principal com menu interativo.
 */
async function main() {
    const host = process.argv[2] || 'localhost';
    const porta = parseInt(process.argv[3]) || 8080;
    
    try {
        const cliente = new ClienteRemotoAPI(host, porta);
        const rl = criarInterface();
        
        console.log('='.repeat(40));
        console.log('Cliente Remoto Smart Home (JavaScript/Node.js)');
        console.log('='.repeat(40));
        console.log(`Conectado a: ${host}:${porta}`);
        
        let continuar = true;
        
        while (continuar) {
            exibirMenu();
            const opcao = await perguntar(rl, 'Escolha uma opcao: ');
            
            try {
                switch (opcao.trim()) {
                    case '1':
                        await listarDispositivosInterativo(cliente);
                        break;
                    case '2':
                        await obterDispositivoInterativo(cliente, rl);
                        break;
                    case '3':
                        await atualizarDispositivoInterativo(cliente, rl);
                        break;
                    case '4':
                        await executarAcaoInterativo(cliente, rl);
                        break;
                    case '5':
                        await listarRotinasInterativo(cliente);
                        break;
                    case '6':
                        await criarRotinaInterativo(cliente, rl);
                        break;
                    case '7':
                        await listarAlertasInterativo(cliente);
                        break;
                    case '8':
                        await obterComodoInterativo(cliente, rl);
                        break;
                    case '0':
                        continuar = false;
                        console.log('\nEncerrando cliente...');
                        break;
                    default:
                        console.log('\n[ERRO] Opcao invalida! Tente novamente.');
                }
            } catch (error) {
                console.log(`\n[ERRO] ${error.message}`);
            }
            
            if (continuar) {
                await perguntar(rl, '\nPressione Enter para continuar...');
            }
        }
        
        rl.close();
        
    } catch (error) {
        console.error(`[ERRO] ${error.message}`);
        console.error(error.stack);
    }
}

// Executa se for chamado diretamente
if (require.main === module) {
    main();
}

module.exports = ClienteRemotoAPI;

File: clientes/python/cliente_remoto.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Cliente Remoto Smart Home - Python
Comunica-se com o servidor Java via API REST HTTP usando JSON.
"""

import json
import urllib.request
import urllib.parse
from typing import List, Dict, Optional, Any


class ClienteRemotoAPI:
    """Cliente remoto que se comunica com o servidor usando API REST HTTP."""
    
    def __init__(self, host: str = "localhost", porta: int = 8080):
        """
        Inicializa o cliente.
        
        Args:
            host: Endereço do servidor
            porta: Porta do servidor
        """
        self.base_url = f"http://{host}:{porta}"
    
    def _fazer_requisicao(self, endpoint: str, method: str = "GET", 
                         dados: Optional[Any] = None) -> Any:
        """
        Executa uma requisição HTTP e retorna a resposta.
        
        Args:
            endpoint: Endpoint da API
            method: Método HTTP (GET, POST)
            dados: Dados para enviar no corpo da requisição
            
        Returns:
            Resposta deserializada
        """
        url = f"{self.base_url}{endpoint}"
        
        # Prepara a requisição
        if dados is not None:
            # Serializa dados para JSON
            json_data = json.dumps(dados, default=str).encode('utf-8')
            req = urllib.request.Request(url, data=json_data, method=method)
            req.add_header('Content-Type', 'application/json')
            req.add_header('Accept', 'application/json')
        else:
            req = urllib.request.Request(url, method=method)
            req.add_header('Accept', 'application/json')
        
        try:
            print(f"[CLIENTE] {method} {endpoint}")
            with urllib.request.urlopen(req) as response:
                # Verifica Content-Type da resposta
                content_type = response.headers.get('Content-Type', '')
                
                # Lê a resposta
                response_data = response.read()
                
                # Se for JSON, deserializa
                if 'application/json' in content_type:
                    try:
                        return json.loads(response_data.decode('utf-8'))
                    except json.JSONDecodeError as e:
                        raise Exception(f"Erro ao decodificar JSON: {e}")
                else:
                    # Se não for JSON, pode ser erro ou dados binários
                    # Tenta decodificar como UTF-8 (pode ser texto de erro)
                    try:
                        texto = response_data.decode('utf-8')
                        # Se conseguir decodificar, pode ser mensagem de erro
                        raise Exception(f"Servidor retornou texto em vez de JSON: {texto}")
                    except UnicodeDecodeError:
                        # Se não conseguir decodificar, são dados binários
                        raise Exception("Servidor retornou dados binarios (serializacao Java) em vez de JSON. " +
                                      "Certifique-se de que o servidor esta configurado para retornar JSON quando " +
                                      "o header Accept: application/json e enviado.")
        except urllib.error.HTTPError as e:
            try:
                error_msg = e.read().decode('utf-8')
            except:
                error_msg = f"Erro HTTP {e.code}"
            raise Exception(f"Erro HTTP {e.code}: {error_msg}")
    
    def listar_dispositivos(self) -> List[Dict]:
        """Lista todos os dispositivos IoT."""
        return self._fazer_requisicao("/api/dispositivos", "GET")
    
    def obter_dispositivo(self, dispositivo_id: str) -> Optional[Dict]:
        """Obtém um dispositivo por ID."""
        return self._fazer_requisicao("/api/dispositivos/obter", "POST", dispositivo_id)
    
    def atualizar_dispositivo(self, dispositivo_id: str, dispositivo: Dict) -> Optional[Dict]:
        """Atualiza um dispositivo."""
        dados = [dispositivo_id, dispositivo]
        return self._fazer_requisicao("/api/dispositivos/atualizar", "POST", dados)
    
    def executar_acao(self, dispositivo_id: str, comando: str) -> Optional[Dict]:
        """Executa uma ação em um dispositivo."""
        dados = [dispositivo_id, comando]
        return self._fazer_requisicao("/api/dispositivos/acao", "POST", dados)
    
    def listar_rotinas(self) -> List[Dict]:
        """Lista todas as rotinas."""
        return self._fazer_requisicao("/api/rotinas", "GET")
    
    def criar_rotina(self, rotina: Dict) -> Optional[Dict]:
        """Cria uma nova rotina."""
        return self._fazer_requisicao("/api/rotinas/criar", "POST", rotina)
    
    def listar_alertas(self) -> List[Dict]:
        """Lista todos os alertas."""
        return self._fazer_requisicao("/api/alertas", "GET")
    
    def obter_comodo(self, nome_comodo: str) -> Optional[Dict]:
        """Obtém um cômodo por nome."""
        return self._fazer_requisicao("/api/comodos/obter", "POST", nome_comodo)


def exibir_menu():
    """Exibe o menu principal."""
    print("\n" + "=" * 40)
    print("MENU PRINCIPAL")
    print("=" * 40)
    print("1. Listar dispositivos")
    print("2. Obter dispositivo por ID")
    print("3. Atualizar dispositivo")
    print("4. Executar acao em dispositivo")
    print("5. Listar rotinas")
    print("6. Criar rotina")
    print("7. Listar alertas")
    print("8. Obter comodo por nome")
    print("0. Sair")
    print("=" * 40)


def listar_dispositivos_interativo(cliente):
    """Lista todos os dispositivos."""
    print("\n--- Listando Dispositivos ---")
    try:
        dispositivos = cliente.listar_dispositivos()
        print(f"Total: {len(dispositivos)} dispositivos\n")
        
        if not dispositivos:
            print("Nenhum dispositivo encontrado.")
        else:
            for i, d in enumerate(dispositivos, 1):
                print(f"{i}. {d.get('nome', 'N/A')}")
                print(f"   ID: {d.get('id', 'N/A')}")
                print(f"   Tipo: {d.get('tipo', 'N/A')}")
                print(f"   Online: {'Sim' if d.get('online') else 'Nao'}")
                print(f"   Comodo: {d.get('comodo', 'N/A')}")
                print()
    except Exception as e:
        print(f"[ERRO] {e}")


def obter_dispositivo_interativo(cliente):
    """Obtém um dispositivo por ID."""
    print("\n--- Obter Dispositivo ---")
    dispositivo_id = input("Digite o ID do dispositivo: ").strip()
    
    if not dispositivo_id:
        print("[ERRO] ID nao pode ser vazio!")
        return
    
    try:
        dispositivo = cliente.obter_dispositivo(dispositivo_id)
        if dispositivo:
            print("\nDispositivo encontrado:")
            print(f"  Nome: {dispositivo.get('nome', 'N/A')}")
            print(f"  ID: {dispositivo.get('id', 'N/A')}")
            print(f"  Tipo: {dispositivo.get('tipo', 'N/A')}")
            print(f"  Online: {'Sim' if dispositivo.get('online') else 'Nao'}")
            print(f"  Comodo: {dispositivo.get('comodo', 'N/A')}")
        else:
            print("\n[ERRO] Dispositivo nao encontrado!")
    except Exception as e:
        print(f"[ERRO] {e}")


def atualizar_dispositivo_interativo(cliente):
    """Atualiza um dispositivo."""
    print("\n--- Atualizar Dispositivo ---")
    dispositivo_id = input("Digite o ID do dispositivo: ").strip()
    
    if not dispositivo_id:
        print("[ERRO] ID nao pode ser vazio!")
        return
    
    try:
        dispositivo = cliente.obter_dispositivo(dispositivo_id)
        if not dispositivo:
            print("[ERRO] Dispositivo nao encontrado!")
            return
        
        print(f"\nDispositivo atual:")
        print(f"  Nome: {dispositivo.get('nome', 'N/A')}")
        print(f"  Online: {'Sim' if dispositivo.get('online') else 'Nao'}")
        
        novo_nome = input("\nNovo nome (Enter para manter): ").strip()
        if novo_nome:
            dispositivo['nome'] = novo_nome
        
        online_str = input("Online? (s/n, Enter para manter): ").strip().lower()
        if online_str == 's':
            dispositivo['online'] = True
        elif online_str == 'n':
            dispositivo['online'] = False
        
        atualizado = cliente.atualizar_dispositivo(dispositivo_id, dispositivo)
        if atualizado:
            print("\n[SUCESSO] Dispositivo atualizado!")
            print(f"  Nome: {atualizado.get('nome', 'N/A')}")
            print(f"  Online: {'Sim' if atualizado.get('online') else 'Nao'}")
        else:
            print("\n[ERRO] Falha ao atualizar dispositivo!")
    except Exception as e:
        print(f"[ERRO] {e}")


def executar_acao_interativo(cliente):
    """Executa uma ação em um dispositivo."""
    print("\n--- Executar Acao ---")
    dispositivo_id = input("Digite o ID do dispositivo: ").strip()
    
    if not dispositivo_id:
        print("[ERRO] ID nao pode ser vazio!")
        return
    
    comando = input("Digite o comando (ligar/desligar): ").strip()
    
    if not comando:
        print("[ERRO] Comando nao pode ser vazio!")
        return
    
    try:
        resultado = cliente.executar_acao(dispositivo_id, comando)
        if resultado:
            print("\n[SUCESSO] Acao executada!")
            print(f"  Dispositivo: {resultado.get('nome', 'N/A')}")
            print(f"  Online: {'Sim' if resultado.get('online') else 'Nao'}")
        else:
            print("\n[ERRO] Falha ao executar acao!")
    except Exception as e:
        print(f"[ERRO] {e}")


def listar_rotinas_interativo(cliente):
    """Lista todas as rotinas."""
    print("\n--- Listando Rotinas ---")
    try:
        rotinas = cliente.listar_rotinas()
        print(f"Total: {len(rotinas)} rotinas\n")
        
        if not rotinas:
            print("Nenhuma rotina encontrada.")
        else:
            for i, r in enumerate(rotinas, 1):
                print(f"{i}. {r.get('nome', 'N/A')}")
                print(f"   ID: {r.get('id', 'N/A')}")
                print(f"   Acoes: {len(r.get('acoes', []))}")
                print()
    except Exception as e:
        print(f"[ERRO] {e}")


def criar_rotina_interativo(cliente):
    """Cria uma nova rotina."""
    print("\n--- Criar Rotina ---")
    nome = input("Nome da rotina: ").strip()
    
    if not nome:
        print("[ERRO] Nome nao pode ser vazio!")
        return
    
    try:
        dispositivos = cliente.listar_dispositivos()
        if not dispositivos:
            print("[ERRO] Nenhum dispositivo disponivel!")
            return
        
        print("\nDispositivos disponiveis:")
        for i, d in enumerate(dispositivos, 1):
            print(f"{i}. {d.get('nome', 'N/A')} ({d.get('id', 'N/A')})")
        
        escolha = input("\nEscolha o numero do dispositivo: ").strip()
        try:
            idx = int(escolha) - 1
            if idx < 0 or idx >= len(dispositivos):
                print("[ERRO] Escolha invalida!")
                return
            dispositivo_id = dispositivos[idx].get('id')
        except ValueError:
            print("[ERRO] Numero invalido!")
            return
        
        comando = input("Comando (ligar/desligar): ").strip()
        
        rotina = {
            'nome': nome,
            'acoes': [{
                'dispositivoId': dispositivo_id,
                'comando': comando,
                'parametros': {}
            }]
        }
        
        criada = cliente.criar_rotina(rotina)
        if criada:
            print("\n[SUCESSO] Rotina criada!")
            print(f"  Nome: {criada.get('nome', 'N/A')}")
            print(f"  ID: {criada.get('id', 'N/A')}")
        else:
            print("\n[ERRO] Falha ao criar rotina!")
    except Exception as e:
        print(f"[ERRO] {e}")


def listar_alertas_interativo(cliente):
    """Lista todos os alertas."""
    print("\n--- Listando Alertas ---")
    try:
        alertas = cliente.listar_alertas()
        print(f"Total: {len(alertas)} alertas\n")
        
        if not alertas:
            print("Nenhum alerta encontrado.")
        else:
            for i, a in enumerate(alertas, 1):
                print(f"{i}. {a.get('titulo', 'N/A')}")
                print(f"   Mensagem: {a.get('mensagem', 'N/A')}")
                print(f"   Comodo: {a.get('comodo', 'N/A')}")
                print()
    except Exception as e:
        print(f"[ERRO] {e}")


def obter_comodo_interativo(cliente):
    """Obtém um cômodo por nome."""
    print("\n--- Obter Comodo ---")
    nome = input("Digite o nome do comodo: ").strip()
    
    if not nome:
        print("[ERRO] Nome nao pode ser vazio!")
        return
    
    try:
        comodo = cliente.obter_comodo(nome)
        if comodo:
            print("\nComodo encontrado:")
            print(f"  Nome: {comodo.get('nome', 'N/A')}")
            dispositivos = comodo.get('dispositivos', [])
            if dispositivos:
                print(f"  Dispositivos: {len(dispositivos)}")
                for d in dispositivos:
                    print(f"    - {d.get('nome', 'N/A')}")
        else:
            print("\n[ERRO] Comodo nao encontrado!")
    except Exception as e:
        print(f"[ERRO] {e}")


def main():
    """Função principal com menu interativo."""
    import sys
    
    host = "localhost"
    porta = 8080
    
    # Permite passar host e porta como argumentos
    if len(sys.argv) >= 2:
        host = sys.argv[1]
    if len(sys.argv) >= 3:
        porta = int(sys.argv[2])
    
    try:
        cliente = ClienteRemotoAPI(host, porta)
        
        print("=" * 40)
        print("Cliente Remoto Smart Home (Python)")
        print("=" * 40)
        print(f"Conectado a: {host}:{porta}")
        
        continuar = True
        
        while continuar:
            exibir_menu()
            opcao = input("Escolha uma opcao: ").strip()
            
            if opcao == "1":
                listar_dispositivos_interativo(cliente)
            elif opcao == "2":
                obter_dispositivo_interativo(cliente)
            elif opcao == "3":
                atualizar_dispositivo_interativo(cliente)
            elif opcao == "4":
                executar_acao_interativo(cliente)
            elif opcao == "5":
                listar_rotinas_interativo(cliente)
            elif opcao == "6":
                criar_rotina_interativo(cliente)
            elif opcao == "7":
                listar_alertas_interativo(cliente)
            elif opcao == "8":
                obter_comodo_interativo(cliente)
            elif opcao == "0":
                continuar = False
                print("\nEncerrando cliente...")
            else:
                print("\n[ERRO] Opcao invalida! Tente novamente.")
            
            if continuar:
                input("\nPressione Enter para continuar...")
                
    except Exception as e:
        print(f"[ERRO] {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()
